docs/specs/tsd-001-deprecation.md

```diff
@@ -0,0 +1,181 @@
+deprecated at the top (e.g. ***DEPRECATED***).
+
+#### Deprecating an option
+
+To deprecate a plugin open, remove the option from the `sample.conf` file and
+add the deprecation information to the structure field in the code. If you for
+for example want to deprecate the `ssl_enabled` option in `inputs.example` you
+should add
+
+```golang
+type Example struct {
+    ...
+    SSLEnabled bool `toml:"ssl_enabled" deprecated:"1.3.0;1.40.0;use 'tls_*' options instead"`
+}
+```
+
+to schedule the setting for removal in version `1.40.0`. The last element of
+the `deprecated` tag is a user-facing notice similar to plugin deprecation.
+
+#### Deprecating an option-value
+
+Sometimes, certain option values become deprecated or superseded by other
+options or values. To deprecate those option values, remove them from
+`sample.conf` and add the deprecation info in the code if the deprecated value
+is *actually used* via
+
+```golang
+func (e *Example) Init() error {
+    ...
+    if e.Mode == "old" {
+        models.PrintOptionDeprecationNotice(telegraf.Warn, "inputs.example", "mode", telegraf.DeprecationInfo{
+            Since:     "1.23.1",
+            RemovalIn: "1.40.0",
+            Notice:    "use 'v1' instead",
+        })
+    }
+    ...
+    return nil
+}
+```
+
+This will show a warning if the deprecated `v1` value is used for the `mode`
+setting in `inputs.example` with a user-facing notice.
+
+### Submit pull-request for removing code
+
+Once the plugin, plugin option or option-value is deprecated, we have to wait
+for the `RemovedIn` release to remove the code. In the examples above, this
+would be version `1.40.0`. After all scheduled bugfix-releases are done, with
+`1.40.0` being the next release, you can create a pull-request to actually
+remove the deprecated code.
+
+Please make sure, you remove the plugin, plugin option or option value and the
+code referencing those. This might also comprise the `all` files of your plugin
+category, test-cases including those of other plugins, README files or other
+documentation. For removed plugins, please keep the deprecation info in
+`deprecations.go` so users can find a reference when switching from a really
+old version.
+
+Make sure you add an `Important Changes` sections to the `CHANGELOG.md` file
+describing the removal with a reference to your PR.
+
+### Time frames and considerations
+
+When deprecating parts of Telegraf, it is important to provide users with enough
+time to migrate to alternative solutions before actually removing those parts.
+
+In general, plugins, plugin options or option values should only be deprecated
+if a suitable alternative exists! In those cases, the deprecations should
+predate the removal by at least one and a half year. In current release terms
+this corresponds to six minor-versions. However, there might be circumstances
+requiring a pro-longed time between deprecation and removal to ensure a smooth
+transition for users.
+
+## Deprecation Framework
```

Comment:
This whole section has pretty much already been explained, right? I'm not sure what this adds?

Reply:
It specifies that there should be a framework doing the warning and then error-and-exit. Should I remove this section?

Reply:
Yeah, I would, otherwise this feels out of place.